File: src/frontend/wms/lib/app/data/model/wms_recebimento_cabecalho_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:wms/app/infra/infra_imports.dart';
import 'package:wms/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';
import 'package:wms/app/data/domain/domain_imports.dart';

class WmsRecebimentoCabecalhoModel {
	int? id;
	int? idWmsAgendamento;
	DateTime? dataRecebimento;
	String? horaInicio;
	String? horaFim;
	int? volumeRecebido;
	double? pesoRecebido;
	String? inconsistencia;
	String? observacao;
	List<WmsRecebimentoDetalheModel>? wmsRecebimentoDetalheModelList;
	WmsAgendamentoModel? wmsAgendamentoModel;

	WmsRecebimentoCabecalhoModel({
		this.id,
		this.idWmsAgendamento,
		this.dataRecebimento,
		this.horaInicio,
		this.horaFim,
		this.volumeRecebido,
		this.pesoRecebido,
		this.inconsistencia,
		this.observacao,
		this.wmsRecebimentoDetalheModelList,
		this.wmsAgendamentoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'data_recebimento',
		'hora_inicio',
		'hora_fim',
		'volume_recebido',
		'peso_recebido',
		'inconsistencia',
		'observacao',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Data Recebimento',
		'Hora Inicio',
		'Hora Fim',
		'Volume Recebido',
		'Peso Recebido',
		'Inconsistencia',
		'Observacao',
	];

	WmsRecebimentoCabecalhoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idWmsAgendamento = jsonData['idWmsAgendamento'];
		dataRecebimento = jsonData['dataRecebimento'] != null ? DateTime.tryParse(jsonData['dataRecebimento']) : null;
		horaInicio = jsonData['horaInicio'];
		horaFim = jsonData['horaFim'];
		volumeRecebido = jsonData['volumeRecebido'];
		pesoRecebido = jsonData['pesoRecebido']?.toDouble();
		inconsistencia = WmsRecebimentoCabecalhoDomain.getInconsistencia(jsonData['inconsistencia']);
		observacao = jsonData['observacao'];
		wmsRecebimentoDetalheModelList = (jsonData['wmsRecebimentoDetalheModelList'] as Iterable?)?.map((m) => WmsRecebimentoDetalheModel.fromJson(m)).toList() ?? [];
		wmsAgendamentoModel = jsonData['wmsAgendamentoModel'] == null ? WmsAgendamentoModel() : WmsAgendamentoModel.fromJson(jsonData['wmsAgendamentoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idWmsAgendamento'] = idWmsAgendamento != 0 ? idWmsAgendamento : null;
		jsonData['dataRecebimento'] = dataRecebimento != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataRecebimento!) : null;
		jsonData['horaInicio'] = Util.removeMask(horaInicio);
		jsonData['horaFim'] = Util.removeMask(horaFim);
		jsonData['volumeRecebido'] = volumeRecebido;
		jsonData['pesoRecebido'] = pesoRecebido;
		jsonData['inconsistencia'] = WmsRecebimentoCabecalhoDomain.setInconsistencia(inconsistencia);
		jsonData['observacao'] = observacao;
		
		var wmsRecebimentoDetalheModelLocalList = []; 
		for (WmsRecebimentoDetalheModel object in wmsRecebimentoDetalheModelList ?? []) { 
			wmsRecebimentoDetalheModelLocalList.add(object.toJson); 
		}
		jsonData['wmsRecebimentoDetalheModelList'] = wmsRecebimentoDetalheModelLocalList;
		jsonData['wmsAgendamentoModel'] = wmsAgendamentoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idWmsAgendamento = plutoRow.cells['idWmsAgendamento']?.value;
		dataRecebimento = Util.stringToDate(plutoRow.cells['dataRecebimento']?.value);
		horaInicio = plutoRow.cells['horaInicio']?.value;
		horaFim = plutoRow.cells['horaFim']?.value;
		volumeRecebido = plutoRow.cells['volumeRecebido']?.value;
		pesoRecebido = plutoRow.cells['pesoRecebido']?.value?.toDouble();
		inconsistencia = plutoRow.cells['inconsistencia']?.value != '' ? plutoRow.cells['inconsistencia']?.value : 'S';
		observacao = plutoRow.cells['observacao']?.value;
		wmsRecebimentoDetalheModelList = [];
		wmsAgendamentoModel = WmsAgendamentoModel();
		wmsAgendamentoModel?.local_operacao = plutoRow.cells['wmsAgendamentoModel']?.value;
	}	

	WmsRecebimentoCabecalhoModel clone() {
		return WmsRecebimentoCabecalhoModel(
			id: id,
			idWmsAgendamento: idWmsAgendamento,
			dataRecebimento: dataRecebimento,
			horaInicio: horaInicio,
			horaFim: horaFim,
			volumeRecebido: volumeRecebido,
			pesoRecebido: pesoRecebido,
			inconsistencia: inconsistencia,
			observacao: observacao,
			wmsRecebimentoDetalheModelList: wmsRecebimentoDetalheModelListClone(wmsRecebimentoDetalheModelList!),
		);			
	}

	wmsRecebimentoDetalheModelListClone(List<WmsRecebimentoDetalheModel> wmsRecebimentoDetalheModelList) { 
		List<WmsRecebimentoDetalheModel> resultList = [];
		for (var wmsRecebimentoDetalheModel in wmsRecebimentoDetalheModelList) {
			resultList.add(
				WmsRecebimentoDetalheModel(
					id: wmsRecebimentoDetalheModel.id,
					idWmsRecebimentoCabecalho: wmsRecebimentoDetalheModel.idWmsRecebimentoCabecalho,
					idProduto: wmsRecebimentoDetalheModel.idProduto,
					quantidadeVolume: wmsRecebimentoDetalheModel.quantidadeVolume,
					quantidadeItemPorVolume: wmsRecebimentoDetalheModel.quantidadeItemPorVolume,
					quantidadeRecebida: wmsRecebimentoDetalheModel.quantidadeRecebida,
					destino: wmsRecebimentoDetalheModel.destino,
				)
			);
		}
		return resultList;
	}

	
}

File: backend/wms/src/model/wms_recebimento_cabecalho_model.py
from src import db
from src.model.wms_recebimento_detalhe_model import WmsRecebimentoDetalheModel
from src.model.wms_agendamento_model import WmsAgendamentoModel


class WmsRecebimentoCabecalhoModel(db.Model):
    __tablename__ = 'wms_recebimento_cabecalho'

    id = db.Column(db.Integer, primary_key=True)
    data_recebimento = db.Column(db.DateTime)
    hora_inicio = db.Column(db.String(8))
    hora_fim = db.Column(db.String(8))
    volume_recebido = db.Column(db.Integer)
    peso_recebido = db.Column(db.Float)
    inconsistencia = db.Column(db.String(1))
    observacao = db.Column(db.Text)
    id_wms_agendamento = db.Column(db.Integer, db.ForeignKey('wms_agendamento.id'))

    wms_recebimento_detalhe_model_list = db.relationship('WmsRecebimentoDetalheModel', lazy='dynamic')
    wms_agendamento_model = db.relationship('WmsAgendamentoModel', foreign_keys=[id_wms_agendamento])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_wms_agendamento = data.get('idWmsAgendamento')
        self.data_recebimento = data.get('dataRecebimento')
        self.hora_inicio = data.get('horaInicio')
        self.hora_fim = data.get('horaFim')
        self.volume_recebido = data.get('volumeRecebido')
        self.peso_recebido = data.get('pesoRecebido')
        self.inconsistencia = data.get('inconsistencia')
        self.observacao = data.get('observacao')

    def serialize(self):
        return {
            'id': self.id,
            'idWmsAgendamento': self.id_wms_agendamento,
            'dataRecebimento': self.data_recebimento.isoformat(),
            'horaInicio': self.hora_inicio,
            'horaFim': self.hora_fim,
            'volumeRecebido': self.volume_recebido,
            'pesoRecebido': self.peso_recebido,
            'inconsistencia': self.inconsistencia,
            'observacao': self.observacao,
            'wmsRecebimentoDetalheModelList': [wms_recebimento_detalhe_model.serialize() for wms_recebimento_detalhe_model in self.wms_recebimento_detalhe_model_list],
            'wmsAgendamentoModel': self.wms_agendamento_model.serialize() if self.wms_agendamento_model else None,
        }